### PR TITLE
Create Lifecycle behavior reference doc

### PR DESCRIPTION
**Pull request by timkimadobe**
## Description

This PR creates a new Lifecycle reference doc that explains in depth the different lifecycle event flows and pitfalls commonly seen when implementing Lifecycle alongside solution extensions that depend on it, like Analytics.

Current proposed location is under Base extensions > MobileCore > Lifecycle

Definitely open to suggestions on formatting, wording, structure, etc. please let me know your thoughts!

The intent is to evolve the document iteratively in the following stages, getting consensus before moving to the next stage:
1. Create the text version of the content with placeholder text diagrams + explanations
2. Create a single diagram to showcase the diagram format and styling
3. Create the diagrams for the rest of the cases

Example of what a diagram would look like:

mermaid-diagram-2024-04-18-145945

Diagram design intent:
1. Making it easy to understand and digest different Lifecycle scenarios visually 
2. Using mermaid.js to create the diagrams, and save them as comments in the document itself - this make is very easy to maintain and update the diagrams quickly using a rendering tool like https://mermaid.live or GitHub markdown itself (mermaid.js is built-in)
3. Diagrams will be placed in doc as screenshots (no built-in diagramming tools to my knowledge)
<s>4. Layout of diagram is top-to-bottom - this is more flexible in terms of accommodating _n_ number of chained events, and is probably more compatible with more screen layouts (ex: mobile, narrow windows, etc)</s>
4. Layout of diagram is left-to-right - to make better use of the page space, and not extend the document too much vertically. Benefit of SVG in this case is that it should be scalable by the user without losing resolution
5. Using brand colors generally
2. Different colors representing different states
    1. Green for good/expected
    2. Red for incorrect/unexpected
    3. blue for neutral, like app events, time window passing, etc.
6. Different shapes/outlines for different events
    1. Rounded rectangles for lifecycle events
    4. Pills for "app events" or "time passing" events (ex: Session timeout window passing, etc)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+**On app subsequent launches**
+
+1. *App launched*
+2. Start
+3. Pause
+4. *App backgrounded*
+5. *Session timeout window passes*
+6. *App launched*
+7. Start <- Start of new session
+8. ...
```
Comment (calebk1) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:45]: What does the "..." mean in this instance?

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+1. No new session detection logic will be run nor will lifecycle metrics will be incremented.
+2. All lifecycle metrics calculated in the shared state dispatched as a consequence reflect the original start time, not any start-after-start timestamps.
+
+This behavior is only ungated after a pause is seen, and the next start will handle session timeout and other lifecycle logic as normal.
+
+### Pause-after-pause
+
+Pause-after-pause causes the Lifecycle extension's pause timestamp in persistence to be updated to the most recent pause seen. There is no gating behavior like in the start-after-start case. The largest consequence of an unintended pause-after-pause is usually stitching together distinct sessions.
+
+**Missing pause, app terminated** (looks the same as a standard crash, but the reason for the missing pause is different)
```
Comment (calebk1) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:73]: I'd put this information under the "Header"-like area.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+5. Start <- Crash
+
+**Missing pause, app still in memory**
+
+1. ...
+2. Start
+3. *App backgrounded* (missing pause) + optionally: *Session timeout window passes*
+4. Start <- **Not** a crash nor a new session, treated as a start after start - emits lifecycle metrics and does not update the start timestamp in persistence
+
+**Missing start** - extended session length error
```
Comment (calebk1) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:88]: Same comment here. I'd put information like this under the header instead.

**Review by calebk1 — CHANGES_REQUESTED** [review]
Some preliminary comments.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+## General behavior
+
+All events that are **not** Lifecycle start or Lifecycle pause **do not** affect lifecycle state or metrics. The starting of MobileCore via `registerExtensions` also **does not** start a Lifecycle session nor affect lifecycle metrics.
+
+## Correct call patterns
+
+There are generally three main flows for a correct implementation.
+
+**On app first launch after fresh install**
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:27]: Reword suggestion, "On app first launch after installation"

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+1. *App launched*
+2. Start <- Start of new session
+3. Pause
+4. Start
+5. Pause
+6. *App backgrounded*
+
+**On app subsequent launches**
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:36]: Suggest rewording, "On subsequent app launches"

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+3. App crashed/force closed - consequences:
+   1. Causes missing pause call
+   2. App removed from device memory
+4. *Session timeout window passes*
+5. *App launched*
+6. Start <- Crash detected
+
+## Incorrect call patterns
+
+### Start-after-start
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:60]: Suggest renaming this phrase to "Consecutive lifecycleStart API calls"

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+Start-after-start causes the Lifecycle extension's shared state version to be updated, but crucially, the start time in persistence **is not** changed. Start-after-start means that:
+
+1. No new session detection logic will be run nor will lifecycle metrics will be incremented.
+2. All lifecycle metrics calculated in the shared state dispatched as a consequence reflect the original start time, not any start-after-start timestamps.
+
+This behavior is only ungated after a pause is seen, and the next start will handle session timeout and other lifecycle logic as normal.
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:67]: I think we could simplify this by staging something similar to:

1. Consecutive lifecycleStart calls are ignored.
2. A consecutive lifecycleStart call will trigger a new shared state event, however the Lifecycle data is not updated from the initial lifecycleStart call and the session start timestamp is unchanged.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+### Pause-after-pause
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:69]: Suggest renaming this phrase to "Consecutive lifecyclePause API calls"

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+Pause-after-pause causes the Lifecycle extension's pause timestamp in persistence to be updated to the most recent pause seen. There is no gating behavior like in the start-after-start case. The largest consequence of an unintended pause-after-pause is usually stitching together distinct sessions.
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:71]: Suggest rewording to something like:
"Consecutive lifecyclePause calls will update the persisted pause timestamp. On the next lifecycleStart API call, the last pause timestamp is used to determine if the previous session has timed out and a new session is started."

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md (hunk 2 above)
Comment (kevinlind) [reply src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:73]: I'd suggest adding these three scenarios under a new "Troubleshooting" header, as @calebk1  suggested.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+This scenario looks the same as a standard crash, but the underlying reason is an app implementation that causes a missing `lifecyclePause` call.
+
+1. ...
+2. `lifecycleStart`
+3. *App backgrounded* (missing pause)
+4. *App removed from device memory* (user or device initated)
+5. `lifecycleStart` <- Crash
+
+**Missing pause, app still in memory**  
+This scenario shows an example of [consecutive `lifecycleStart` API calls](#consecutive-lifecyclestart-api-calls).
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:85]: Add a little more to the description: "This scenarios shows an example of consecutive `lifecycleStart` API calls. The app close is not detected because `lifecyclePause` is not called. As the app is not removed from memory, the current session continues."

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+1. ...
+2. `lifecycleStart`
+3. *App backgrounded* (missing pause) + optionally: *Session timeout window passes*
+4. `lifecycleStart`
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:90]: Add a line for "App reopended" or "App foregrounded"

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+3. *App backgrounded* (missing pause)
+4. *App removed from device memory* (user or device initated)
+5. `lifecycleStart` <- Crash
+
+**Missing pause, app still in memory**  
+This scenario shows an example of [consecutive `lifecycleStart` API calls](#consecutive-lifecyclestart-api-calls).
+
+1. ...
+2. `lifecycleStart`
+3. *App backgrounded* (missing pause) + optionally: *Session timeout window passes*
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:89]: change to "missing `lifecyclePause` call". Also, remove the "optionally: Session timeout window passes" as it is extraneous information to this scenario.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+    classDef regularBox fill:#009c3b,stroke:#009c3b,color:#fff;
+    classDef incorrectBox fill:#EB1000,stroke:#EB1000,color:#fff;
+
+    class A,D regularPill;
+    class B,C regularBox;
+-->
+
+**Subsequent app launches, continue session with app close**
+
+1. Expected Lifecycle event(s)
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:68]: It's not clear to me what "Expected Lifecycle event(s)" is. If this is "App Launch" followed by "lifecycleStart" maybe state that/include in diagram instead?

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+## Expected Lifecycle scenarios
+
+The following Lifecycle workflows show examples of the expected ordering of Lifecycle events.
+
+**App first launch after installation**
+
+1. *App launched*
+2. `lifecycleStart` (Start of new session)
+3. `lifecyclePause`
+4. *App closed by user*
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:46]: For each scenario, are you planning on removing the list of steps and just keep the diagrams or are you planning on keeping both?  With the diagrams, I don't see a need to repeat the information in a list as well.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+1. Expected Lifecycle event(s)
+2. `lifecycleStart`
+3. *App closed by user (missing `lifecyclePause` call)*
+4. *App terminated*
+5. `lifecycleStart` (Crash reported)
+
+![](./assets/android/lifecycle-missing-pause-terminated.svg)
+
+**Missing pause, app still in memory**  
+This scenarios shows an example of [consecutive `lifecycleStart` API calls](#consecutive-lifecyclestart-api-calls). The app close is not detected because `lifecyclePause` is not called. As the app is not removed from memory, the current session continues.
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:197]: "This scenario ...", (remove 's')

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 16]
```diff
@@ -226,33 +204,27 @@ In this scenario, the `lifecycleStart` call (5) is not detected as a new session
...
-    class C,E regularPill;
-    class D dashedPill;
-    class F incorrectBox;
+    class A,D,F regularPill;
+    class B,C regularBox;
+    class E dashedPill;
+    class G incorrectBox;
 -->
 
 In this scenario, the last `lifecyclePause` call (6) is effectively a [consecutive `lifecyclePause` API call](#consecutive-lifecyclepause-api-calls), with the consequences of:
```
Comment (kevinlind) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:230]: Update the reference numbers here. The lifecyclePause calls are at 3 and 7, not 2 and 6.

**Review by kevinlind — COMMENTED** [review]
Getting close. One more change.

**Review by kevinlind — APPROVED** [review]

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+    class A,C,D regularPill;
+    class B,E regularBox;
+-->
+
+## Troubleshooting unexpected Lifecycle scenarios
+
+The following Lifecycle workflows show examples of the unexpected ordering of Lifecycle events that can impact Lifecycle data and can indicate an incorrect implementation of the Lifecycle APIs. To address the following scenarios, refer to the [implementation guide for Lifecycle](index.md#register-lifecycle-with-mobile-core-and-add-appropriate-startpause-calls).
+
+**Missing pause, app terminated**  
+This scenario looks the same as a standard crash, but the underlying reason is an incorrect implementation where `lifecyclePause` is not called before the app is terminated.
```
Comment (calebk1) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:157]: Missing newline between the text and the bolded items.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+    classDef regularBox fill:#009c3b,stroke:#009c3b,color:#fff;
+    classDef incorrectBox fill:#EB1000,stroke:#EB1000,color:#fff;
+
+    class A,D regularPill;
+    class B,E regularBox;
+    class C incorrectBox;
+-->
+
+**Missing pause, app still in memory**  
+This scenario shows an example of [consecutive `lifecycleStart` API calls](#consecutive-lifecyclestart-api-calls). The app close is not detected because `lifecyclePause` is not called. As the app is not removed from memory, the current session continues.
```
Comment (calebk1) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:180]: Same here - missing newline.

File: src/pages/home/base/mobile-core/lifecycle/behavior-reference.md [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+
+    class A,D,E regularPill;
+    class B,E regularBox;
+    class C incorrectBox;
+-->
+
+In this scenario, the `lifecycleStart` call (5) is not detected as a new session nor a crash. It will have the consequences of a [consecutive `lifecycleStart` API call](#consecutive-lifecyclestart-api-calls).
+
+**Missing start**  
+This scenario shows an example of [consecutive `lifecyclePause` API calls](#consecutive-lifecyclepause-api-calls). The new session is not detected because `lifecycleStart` is not called.
```
Comment (calebk1) [inline src/pages/home/base/mobile-core/lifecycle/behavior-reference.md:205]: Same here - missing a newline.

**Review by calebk1 — COMMENTED** [review]
Looks good!

A few comments below, but also - you'll need to add this to the gatsby-config.js in order to get this new doc render.
